File: src/gtm_sourcing_agent/stages/talent_map.py
"""Stage 4-5: Talent Market Mapping + Title Intelligence. Populates
target_companies and title_intelligence on the TalentMap; search
strategies are added separately by stages/search_strategy.py so the two
can be regenerated independently. See prompts/talent_map.md."""

import json

from .. import llm_client, storage
from ..models import TalentMap

MIN_TARGET_COMPANIES = 15  # the prompt asks for >=5 per tier across 3 tiers


def run(role_id: str, *, storage_backend=storage) -> TalentMap:
    icp = storage_backend.require_section(role_id, "icp")
    prompt = llm_client.render_prompt("talent_map.md", icp_json=json.dumps(icp))
    result = llm_client.generate(prompt, TalentMap, stage="talent_map")

    if len(result.target_companies) < MIN_TARGET_COMPANIES:
        # One retry, with the shortfall spelled out — a real safety net
        # for thin coverage, not a way to pad the list with invented
        # companies. If the retry still comes up short, we keep it and
        # let the count speak for itself.
        total = len(result.target_companies)
        retry_prompt = prompt + (
            f"\n\nYour previous attempt returned only {total} target companies total — this role needs "
            f"at least {MIN_TARGET_COMPANIES} (5+ per tier). Cover more real companies across all three "
            f"tiers and try again."
        )
        retried = llm_client.generate(retry_prompt, TalentMap, stage="talent_map")
        if len(retried.target_companies) > total:
            result = retried

    existing = storage_backend.load_role(role_id).get("talent_map") or {}
    result.search_strategies = existing.get("search_strategies", []) or result.search_strategies
    storage_backend.merge_section(role_id, "talent_map", result.model_dump())
    return result

File: backend-node/src/stages/talentMap.ts
// Port of stages/talent_map.py -- Stage 4-5: Talent Market Mapping +
// Title Intelligence. Populates target_companies/title_intelligence;
// search_strategies is added separately by searchStrategy.ts so the two
// regenerate independently.
import * as storage from "../db/storage.js";
import * as llmClient from "../llmClient.js";
import { TalentMap } from "../models.js";

const MIN_TARGET_COMPANIES = 15; // the prompt asks for >=5 per tier across 3 tiers

export async function run(roleId: string): Promise<TalentMap> {
  const icp = await storage.requireSection(roleId, "icp");
  const prompt = llmClient.renderPrompt("talent_map.md", { icp_json: JSON.stringify(icp) });
  let result = await llmClient.generate(prompt, TalentMap, { stage: "talent_map" });

  if (result.target_companies.length < MIN_TARGET_COMPANIES) {
    const total = result.target_companies.length;
    const retryPrompt = prompt +
      `\n\nYour previous attempt returned only ${total} target companies total — this role needs ` +
      `at least ${MIN_TARGET_COMPANIES} (5+ per tier). Cover more real companies across all three ` +
      `tiers and try again.`;
    const retried = await llmClient.generate(retryPrompt, TalentMap, { stage: "talent_map" });
    if (retried.target_companies.length > total) result = retried;
  }

  const existing = (await storage.loadRole(roleId)).talent_map ?? {};
  result.search_strategies = existing.search_strategies?.length ? existing.search_strategies : result.search_strategies;
  await storage.mergeSection(roleId, "talent_map", result);
  return result;
}
